music_assistant/providers/audible/__init__.py

```diff
@@ -0,0 +1,300 @@
+from music_assistant_models.errors import LoginFailed
+
+from music_assistant.models.music_provider import MusicProvider
+from music_assistant.providers.audible.audible_helper import (
+    AudibleHelper,
+    audible_custom_login,
+    audible_get_auth_info,
+    check_file_exists,
+    remove_file,
+)
+
+if TYPE_CHECKING:
+    from music_assistant_models.media_items import Audiobook
+    from music_assistant_models.provider import ProviderManifest
+    from music_assistant_models.streamdetails import StreamDetails
+
+    from music_assistant.mass import MusicAssistant
+    from music_assistant.models import ProviderInstanceType
+
+
+# Constants for config actions
+CONF_ACTION_AUTH = "authenticate"
+CONF_ACTION_VERIFY = "verify_link"
+CONF_ACTION_CLEAR_AUTH = "clear_auth"
+CONF_AUTH_FILE = "auth_file"
+CONF_POST_LOGIN_URL = "post_login_url"
+CONF_CODE_VERIFIER = "code_verifier"
+CONF_SERIAL = "serial"
+CONF_LOGIN_URL = "login_url"
+CONF_LOCALE = "locale"
+
+
+async def setup(
+    mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
+) -> ProviderInstanceType:
+    """Initialize provider(instance) with given configuration."""
+    return Audibleprovider(mass, manifest, config)
+
+
+async def get_config_entries(
+    mass: MusicAssistant,
+    instance_id: str | None = None,  # noqa: ARG001
+    action: str | None = None,
+    values: dict[str, ConfigValueType] | None = None,
+) -> tuple[ConfigEntry, ...]:
+    """
+    Return Config entries to setup this provider.
+
+    instance_id: id of an existing provider instance (None if new instance setup).
+    action: [optional] action key called from config entries UI.
+    values: the (intermediate) raw values for config entries sent with the action.
+    """
+    if values is None:
+        values = {}
+
+    locale = cast(str, values.get("locale", "") or "us")
+    auth_file = cast(str, values.get(CONF_AUTH_FILE))
+
+    # Check if auth file exists and is valid
+    auth_required = True
+    if auth_file and await check_file_exists(auth_file):
+        try:
+            auth = await asyncio.to_thread(audible.Authenticator.from_file, auth_file)
+            auth_required = False
+        except Exception:
+            auth_required = True
+    label_text = (
+        (
+            "You need to authenticate with Audible. Click the authenticate button below "
+            "to start the authentication process which will open in a new (popup) window, "
+            "so make sure to disable any popup blockers.\n\n"
+            "NOTE: \n"
+            "After successful login you will get a 'page not found' message - this is expected. "
+            "Copy the address to the textbox below and press verify. "
+            "This will register this provider as a virtual device with Audible."
+        )
+        if auth_required
+        else (
+            "Successfully authenticated with Audible."
+            "\nNote: Changing marketplace needs new authorization"
+        )
+    )
```

Comment:
Nit: This has become a bit unreadable. Maybe turn it into a regular if..else clause ?